### ISSUE TITLE
Soundness bug in z3str3

### ISSUE THREAD

**Issue by wintered** [Z3Prover/z3#4306]
```
[549] % z3release small.smt2
unsat
[550] % z3release smt.string_solver=z3str3 small.smt2
sat
[551] % cat small.smt2
(declare-fun a () String)
(assert (str.in.re a (re.inter (re.opt re.allchar) (str.to.re "ab"))))
(check-sat-using ctx-solver-simplify)
[552] %
```

OS: Ubuntu 18.04
Commit: 32055a3

**Comment by zhendongsu** [comment]
Hi Murphy and Nikolaj, we'll be filing bug reports in z3str3 unless you instruct us to hold off testing z3str3.  We'll start with a few first.  Thank you. 

@NikolajBjorner @mtrberzi

**Comment by NikolajBjorner** [comment]
It is of course great that you are spending the effort to test z3str3.

My main reservation is that these bugs are not being fixed at a rate that comes 
close to the filing rate and many old bugs that should be possible to deal 
with have not been addressed. An effect is that the github issue tracker gets clobbered with z3str3 bugs and I end up not being effective in identifying actionable bugs elsewhere. 
I prefer that bugs be coalesced.

In the limit, if bugs remain without being addressed in a component I will just remove it to not deal with the liability. For example, I removed interpolation features because bugs were not being addressed. Other examples of removed components are smaller, such as unit_walk sat, bit-width reduction.

**Comment by zhendongsu** [comment]
That makes much sense, Nikolaj; thanks for your candid thoughts.  

What do you think, Murphy?   @mtrberzi

**Comment by mtrberzi** [comment]
I agree, if it is possible to coalesce bugs I would prefer that as well, especially if the instances are very similar. All reported issues will be addressed as my own time and ability allows.

**Comment by wintered** [comment]
Seems fixed (not by 5a9b0dd though)

```
$z3release smt.string_solver=z3str3 bug.smt2
unsat
$z3release smt.string_solver=z3str3 bug.smt2
unsat
$cat bug.smt2 
(declare-fun a () String)
(assert (str.in.re a (re.inter (re.opt re.allchar) (str.to.re "ab"))))
(check-sat-using ctx-solver-simplify)
```
